lib/module-entry.js

```diff
@@ -0,0 +1,227 @@
+'use strict';
+
+var heimdall = require('heimdalljs');
+var genNameDirKey = require('./gen-name-dir-key');
+var resolvePackagePath = require('./resolve-package-path');
+var cacheKey = require('./cache-key');
+var tryRequire = require('./try-require');
+var crypto = require('crypto');
+
+/*
+ * Entries in 'MODULE_ENTRY' cache. The idea is to minimize the information needed for
+ * each entry (don't store the entire package.json object), but keep
+ * information around for all parts of computing a hash for the dependency
+ * tree of the original package named by 'name' and 'dir' in hashForDep.
+ * The ModuleEntry is created at the time the package.json file is
+ * read in (so we only have to do that once), but some properties
+ * may not be initialized until later (this then supports creating cycles
+ * in the dependencies list, which is supported by Node).
+ */
+module.exports = ModuleEntry;
+function ModuleEntry(name, version, rootDir, sourceHash) {
+
+  // the resolved real file path for the package.json file,
+  // the same value as 'realPath' in genRealPathKey.
+  this.rootDir = rootDir;
+
+  // the name of the package, from its package.json
+  this.name = name;
+
+  // the version of the package, from its package.json
+  this.version = version;
+
+  // the computed hash for only the package's source files (under
+  // the root dir, but specifically NOT any node_modules dependencies).
+  this._sourceHash = sourceHash;
+
+  // the computed hash for the package's source files plus
+  // the source hashes of the dependencies. Because we compute
+  // dependencies as 'all the way down', this gets everything.
+  this._hash = null;
+
+  // references to other ModuleEntry objects that correspond
+  // to the dependencies listed in the package.json
+  this._dependencies = null;
+}
+
+ModuleEntry.prototype.initializeDependencies = function(caches, pkg, hashTreeFn) {
+  if (this._dependencies !== null) {
+    throw new Error('[hashForDep] Cannot re-initializeDependencies a moduleEntry');
+  }
+  this._dependencies = Object.create(null);
+
+  if (!pkg.dependencies) { return; }
+
+  var moduleEntry = this;
+  var Constructor = this.constructor;
+  // Recursively locate the references to other moduleEntry for the package's
+  // dependencies. Initially we just do the 'local' dependencies (i.e. the ones referenced
+  // in the package.dependencies field.) Later, hashForDep calls 'gatherDependencies'
+  // to compute the complete list including transitive dependencies, and that is then
+  // used for the final hash calculation.
+  Object.keys(pkg.dependencies).sort().forEach(function(dep) {
```

Comment:
Should probably also care about `optionalDependencies`, right?

Reply:
The original code didn't specifically mention optionalDependencies (I'd never heard of that property, just dependencies and devDependencies), but if they should be tracked for changes, agree with your point. Do you agree given the above comment?